processing/src/main/java/io/druid/query/MultiSourceBaseQuery.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Ordering;
+import io.druid.java.util.common.guava.Sequence;
+
+import java.util.Map;
+
+public abstract class MultiSourceBaseQuery<T extends Comparable<T>> implements Query<T>
```

Comment:
Well, I think this is needed for future extension. It currently has only a single subclass, but can have other ones in the future. For example, we can add more classes for different join algorithms like we have several classes (groupBy, topN, and timeseries) for aggregation queries.